### DISCUSSION TITLE
CP-SAT: Int/Boolean Constraint

### DISCUSSION THREAD

**Discussion by ngaihiang** [google/or-tools#3208] [CP-SAT questions] [answered]
Hi, I am having trouble modelling a constraint. 

I have a list of boolean variables - x0, x1, ...x49(~50 variables)

I want to have another variable to track the sum of the list of boolean variables: y = sum(x1, x2, ...x50)

And if the sum of these 50 exceeds 20 (y>20), then I will set another new boolean variable z = true

I tried something like this:

`
xx = dict()
for i in range(50):
----xx[i] = model.NewBoolVar('xx%i' % (i))

b = model.NewBoolVar('b') # intermediate variable
z = model.NewBoolVar('z') # intermediate variable

#### using b as intermediate variable for channeling
model.Add(sum(xx[i] for i in range(50)) >20).OnlyEnforceIf(b)
model.Add(sum(xx[i] for i in range(50)) <=20).OnlyEnforceIf(b.Not())

model.Add(z==1).OnlyEnforceIf(b)
model.Add(z==0).OnlyEnforceIf(b.Not)`

I didnt include y variable here but used b as an intermediate variable like what I read here: 
https://developers.google.com/optimization/cp/channeling#:~:text=A%20channeling%20constraint%20links%20variables,variable%20to%20a%20particular%20value%22.

However, I am getting this error 
` raise TypeError('Not supported: CpModel.Add(' + str(ct) + ')')
TypeError: Not supported: CpModel.Add([<ortools.sat.python.cp_model.BoundedLinearExpression object at 0x7fc7fe789150>])`

Is this the right way to create a variable based on other boolean variables?

Appreciate any help !!

**Comment by lperron** [comment, marked as answer]
model = cp_model.CpModel()
        xx = dict()
        for i in range(50):
            xx[i] = model.NewBoolVar('xx%i' % (i))

        b = model.NewBoolVar('b') # intermediate variable
        z = model.NewBoolVar('z') # intermediate variable

        # b implies sum of list is >20
        model.Add(sum(xx[i] for i in range(50)) > 20).OnlyEnforceIf(b)
        model.Add(sum(xx[i] for i in range(50)) <= 20).OnlyEnforceIf(b.Not())

        model.Add(z == 1).OnlyEnforceIf(b)
        model.Add(z == 0).OnlyEnforceIf(b.Not())

works flawlessly

you forgot the () after the last Not

and in  you example, you can simple set model.Add(z == b), or even better only use one variable.
